pkg/simple/client/logging/elasticsearch/elasticsearch.go

```diff
@@ -34,76 +40,102 @@ type client interface {
 }
 
 func NewElasticsearch(options *Options) (*Elasticsearch, error) {
-	var version, index string
-	es := &Elasticsearch{}
-
-	if options.Version == "" {
-		var err error
-		version, err = detectVersionMajor(options.Host)
-		if err != nil {
-			return nil, err
-		}
-	} else {
-		version = options.Version
-	}
-
-	if options.IndexPrefix != "" {
-		index = options.IndexPrefix
-	} else {
-		index = "logstash"
+	var err error
+	es := &Elasticsearch{
+		host:    options.Host,
+		version: options.Version,
+		index:   options.IndexPrefix,
 	}
 
-	switch version {
+	switch es.version {
 	case ElasticV5:
-		es.c = v5.New(options.Host, index)
+		es.c, err = v5.New(es.host, es.index)
 	case ElasticV6:
-		es.c = v6.New(options.Host, index)
+		es.c, err = v6.New(es.host, es.index)
 	case ElasticV7:
-		es.c = v7.New(options.Host, index)
+		es.c, err = v7.New(es.host, es.index)
+	case "":
+		es.c = nil
 	default:
-		return nil, fmt.Errorf("unsupported elasticsearch version %s", version)
+		return nil, fmt.Errorf("unsupported elasticsearch version %s", es.version)
 	}
 
-	return es, nil
+	return es, err
 }
 
-func (es *Elasticsearch) ES() *client {
-	return &es.c
+func (es *Elasticsearch) getClient() client {
+	es.mux.RLock()
```

Comment:
This is a typical singleton with lazy load,  there is a great post http://marcio.io/2015/07/singleton-pattern-in-go/

Reply:
I have seen these kind of posts around. The problem is that `sync.Once` performs exactly once and no more. What if the init task fails?

Reply:
OK, I see.

Reply:
It's better to initialize client in one place, such as
```suggestion
	func (es *Elasticsearch) getClient() (client, error) {
	  if es.c != nil {
	    return es.c, nil
	  }
	  
	  es.mux.Lock()
	  defer es.mu.Unlock()
	    err = es.initClient()
	    if err == nil && es.c != nil {
	      return es.c, nil
	    }
	    
	    return nil, err
	  
	}
```

Reply:
to be more clear, maybe we could add a else branch:
```
	  if es.c != nil {
	    return es.c, nil
	  } else {
	    es.mux.Lock()
	    defer es.mu.Unlock()
	    err = es.initClient()
	    if err == nil && es.c != nil {
	      return es.c, nil
	    }
	    return nil, err
	  }
	}
```

Reply:
I put code comments to clarify this issue.

Reply:
I put code comments to clarify this issue.

Reply:
@huanggze The new change looks better and pls ignore my previous code snippet, defer is only call at the end of a function, not the end of a scope